### ISSUE TITLE
slow formatting of fairly trivial file

### ISSUE THREAD

**Issue by gdanov** [clojure-lsp/clojure-lsp#266]
while triaging [this issue](https://github.com/emacs-lsp/lsp-mode/issues/2537#issuecomment-762843091) we saw clojure-lsp needs half a second for formatting 2k lines of identical & very trivial expression. This is slow according to my *expectations* ... I'm not aware of benchmark or something that sets the baseline what's 'normal'. If this is the best performance achievable maybe something must be done on lsp-mode side to mitigate.

I'm creating this issue as requested by @ericdallo to track the potential performance problem.

**Comment by ericdallo** [comment]
Thanks for the report!
Clojure-lsp just leave the format to cljfmt, do you have a sample file where you notice the slowliness?
I suspect this happen with huge forms/functions, if you have the same file with multiple not big functions this will be fast

**Comment by ericdallo** [comment]
@gdanov, I may found the issue, but have no idea how we could improve that, the issue is that when clojure-lsp format a range, we can't format only the range as we don't know if that range is inside a function or any form, so clojure-lsp format the whole form, example:
```clojure
(def a 1)

(defn foo []
   (+ 1 2))

(def b 2)
```
with that code, if you try to range format only `(+ 1 2)`, clojure-lsp will format the whole `foo` function.
In your case, as `(+ (+ 1 2 3))` is not inside a form, we try to format the whole document

**Comment by gdanov** [comment]
How about not formatting it after paredit change?

**Comment by ericdallo** [comment]
this is not clojure-lsp responsibility, probably would need to be done on lsp-mode if possible

**Comment by snoe** [comment]
we may be able to index form ranges with rewrite-clj when files open instead of when formatting. Alternatively it might be a useful function to clj-format to (format-form-at line col)

**Comment by ericdallo** [comment]
I made some tests and what seems to be the issue is the need to format the whole parent form instead of the selected one.
For example, formatting:
```clojure
(a 
  (b
   |(c
     (+ 2 3))|))
```

With the | as the selected region, it will format the whole `a` macro, since formatting only the selected region would result in wrong spaces since cljfmt doesn't know the spacing before the `c` macro.

This behavior on a big function/form/test will result sometimes in a slowness on cljfmt side even if user selected a small region of code to format or in this case, using paredit, where call lsp-mode ident to format after barf/slurp operations.  So not sure how to fix that

**Comment by snoe** [comment]
We could probably hack something here.

If the `      (c ...)` form is to be indented and starts on col 6 then what you get back should be able to add 6 spaces to the start of each subsequent lines

```clojure
   (a |(c ...
       (indent-me ... ))

=> ; (from clj-format)
(c ...
  (indent-me ... ))

=> ; (after adding indents)
(c ...
         (indent-me ... ))
```
Then our edit would be from the start to the end of the `(c ...)` form

**Comment by ericdallo** [comment]
Looks a valid approach, not so easy, but it's possible indeed

**Comment by gdanov** [comment]
I use paredit with cider's formatting and never noticed slowness. I have no idea how it's formatting works, but maybe there's some strategy that could be borrowed

**Comment by ericdallo** [comment]
Yeah, I gave a try on @snoe idea but failed :/, could not calculate correctly the spaces to add, a fix/option on cljfmt could be a good approach too.

**Comment by ericdallo** [comment]
I created this issue on cljfmt side: https://github.com/weavejester/cljfmt/issues/212

**Comment by ericdallo** [comment]
I think this was fixed on master finally @gdanov :) It was a performance issue on cljfmt fixed: https://github.com/weavejester/cljfmt/issues/226

It'd be great if you could confirm the improvements, it works great on my side.
I should release in a couple of days

**Comment by gdanov** [comment]
Thanks! Do I guess correctly that I need to be using latest master clojure-lsp ? no need to build cljfmt

**Comment by ericdallo** [comment]
Yes @gdanov or wait for next clojure-lsp release

**Comment by gdanov** [comment]
Testing report: using rev /HEAD-f3f975c (via brew) is still significantly slower than cider/clojure mode reformatting. I took into account that the sexp must not be top-level, but this does not make any difference.

Reverting to the latest brew release (20210601) makes the performance ~2 times worse than the latest release so there is some improvement.

Cljfmt is not my preferred formatting tool so I wouldn't fight for this to be re-opened.

**Comment by ericdallo** [comment]
@gdanov yeah, for huge buffers this is still noticable, but it was a great improvement, maybe @LuisThiamNye knows if there is more to be improved on cljfmt side?

**Comment by LuisThiamNye** [comment]
I would expect there to be micro-performance improvements to be had in clojure-lsp, cljfmt, and rewrite-clj. Perhaps [clj-fast](https://github.com/bsless/clj-fast) and [Naked Performance (with Clojure) – Tommi Reiman - YouTube](https://www.youtube.com/watch?v=3SSHjKT3ZmA) could be an inspiration.

The tricky thing is that there are no individual operations that stand out to be consuming a particularly large amount of time. Rather, it is an accumulation of an incredible number of small operations that contribute to slowness.

In addition to refining frequently used functions for performance, it might also be worth considering the architecture of cljfmt as a whole to see if it could be redesigned for efficiency.

**Comment by ryfow** [comment]
I know this ticket is old, but I thought I'd just drop a note here that if you're running into exceptionally slow smartparens performance, try upgrading to the latest & greatest clojure-lsp before wandering down too many other rabbit holes.

**Comment by adityaathalye** [comment]
_NOTE: Cross-posting a related Clojurians Slack discussion I had yesterday. No action needed. This is for reference only._

TL;DR: As mentioned by ericdallo in that thread, it is a known limitation. The solution is to disable range formatting. (I have set `lsp-enable-indentation` to `nil`).

---

The [discussion](https://clojurians-log.clojureverse.org/lsp/2024-08-04/1722763198.699399):

High CPU usage from the clojure-lsp server was plaguing me in the malli codebase (especially the malli.core namespace, which is ~2,700 lines of code).

Indentation behaviour causes the CPU spike when I slurp/barf forms (at any level of nesting) using smartparens (e.g. sp-forward-barf-sexp and sp-forward-slurp-sexp).
- Workspace log shows that the client sends the entire file's worth of content (~2,700 lines of code) to the LSP server, for any slurp/barf action. This is happening to provide clojure-lsp-powered re-indentation.
```
[Trace - 02:28:42 PM] Sending notification 'textDocument/didChange'.
Params: {
  "textDocument": {
   "uri": "file:///path/to/malli/src/malli/core.cljc",
   "version": 30
  },  
   "contentChanges": [
        {"text": "(ns malli.core\n

... [SNIP] ...

the whole file worth of content... ... ..."
 }
]
```
- I assume that's a bit much for cljfmt, because the rangeFormatting request times out.
```
lsp-request: Timeout while waiting for response.  Method: textDocument/rangeFormatting
```
- Meanwhile, my Emacs remains frozen for several seconds (until the timeout).
- My Emacs becomes responsive if I turn off lsp indentation support by setting lsp-enable-indentation nil so that Emacs uses CIDER indentation.

I am on Emacs 28.1, Ubuntu 28.01, Lenovo Thinkpad E460, using the latest available versions for all of these:
- smartparens: 20240713.1002
- lsp-mode: 20240801.2341 
- clojure-mode: 20240526.1825
- clojure-lsp 2024.04.22-11.50.26
- clj-kondo 2024.03.13
